File: Files, Directories and Exceptions/Exercises/Book Library/Book Library.cs
using System;
using System.Linq;
using System.Text;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using static System.Console;

class Book
{

    public string Title { set; get; }
    public string Publisher { set; get; }
    public string ReleaseDate { set; get; }
    string ISBN { set; get; }
    public decimal Price { set; get; }

    public Book(string title, string publisher, string releaseDate, string ISBN, string price)
    {
        Title = title;
        Publisher = publisher;
        ReleaseDate = releaseDate;
        this.ISBN = ISBN;
        Price = decimal.Parse(price);
    }
}

class Library
{
    Dictionary<string, List<Book>> authorBooks = new Dictionary<string, List<Book>>();

    public void AddAuthor(string name, Book book) => authorBooks[name] = new List<Book> { book };

    public void AddBook(string authorName, Book book) => authorBooks[authorName].Add(book);

    public bool HasAuthor(string name) => authorBooks.ContainsKey(name);

    public string ListAuthorSalery()
    {
        var authorSalery = authorBooks.ToDictionary(k => k.Key, v => v.Value.Sum(a => a.Price));
        authorSalery = authorSalery.OrderByDescending(p => p.Value).ThenBy(a => a.Key)
            .ToDictionary(k => k.Key, v => v.Value);

        var sB = new StringBuilder();
        foreach (var author in authorSalery)
            sB.AppendLine(author.Key + " -> " + author.Value.ToString("f2"));

        return sB.ToString();
    }
}

class Program
{
    static void Main()
    {
        var inText = File.ReadAllLines(@"X:\Development\SoftUni\Files, Directories and Exceptions\Exercises\Book Library\input.txt");
        var lib = new Library();

        for (int i = 1; i < int.Parse(inText[0])+1; i++)
        {
            var input = inText[i].Split().ToArray();
            string bookName = input[0];
            string bookAuthor = input[1];
            string publisher = input[2];
            string releaseDate = input[3];
            string ISBN = input[4];
            string price = input[5];

            if (lib.HasAuthor(bookAuthor))
                lib.AddBook(bookAuthor, new Book(bookName, publisher, releaseDate, ISBN, price));
            else
                lib.AddAuthor(bookAuthor, new Book(bookName, publisher, releaseDate, ISBN, price));
        }

        File.WriteAllText(@"X:\Development\SoftUni\Files, Directories and Exceptions\Exercises\Book Library\output.txt",lib.ListAuthorSalery());
    }
}

File: Software Technologies/Java/Exercises/Book Library.java
package com.company;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.lang.System.in;

public class Main {
    static Scanner Input = new Scanner(in);

    public static void main(String[] args) {
        Library lib = new Library();

        for (int i = Integer.parseInt(Input.nextLine()); i > 0; i--)
        {
            String[] input     = Input.nextLine().split(" ");
            String bookName    = input[0];
            String bookAuthor  = input[1];
            String publisher   = input[2];
            String releaseDate = input[3];
            String ISBN        = input[4];
            String price       = input[5];

            if (lib.HasAuthor(bookAuthor))
                lib.AddBook(bookAuthor, new Book(bookName, publisher, releaseDate, ISBN, price));
            else
                lib.AddAuthor(bookAuthor, new Book(bookName, publisher, releaseDate, ISBN, price));
        }

        System.out.println(lib.ListAuthorSalery());
    }
}

class Book
{

    public String Title;
    public String Publisher;
    public String ReleaseDate;
    String ISBN;
    public Double Price;

    public Book(String title, String publisher, String releaseDate, String ISBN, String price)
    {
        Title = title;
        Publisher = publisher;
        ReleaseDate = releaseDate;
        this.ISBN = ISBN;
        Price = Double.parseDouble(price);
    }
}

class Author{
    public String Name;
    public ArrayList<Book> Books = new ArrayList<>();

    public  Author(String name, Book book){
        Name=name;
        Books.add(book);
    }

    public Double getSallery() {
        Double sum = 0.00;
        for(Book book:Books)
            sum+=book.Price;
        return sum;
    }
}

class Library
{
    private List<Author> authors = new ArrayList<>();

    public void AddAuthor(String name, Book book){
        authors.add(new Author(name,book));
    }

    public void AddBook(String authorName, Book book) {
        for(Author author: authors)
            if(author.Name.compareTo(authorName)==0)
                author.Books.add(book);
    }

    public Boolean HasAuthor(String name) {
        for(Author author:authors)
            if(author.Name.compareTo(name)==0)
                return true;
        return false;
    }

    public String ListAuthorSalery()
    {
        Comparator<Author> orderBySaleryThenByName = Comparator.comparing(a->a.getSallery(),Comparator.reverseOrder());
        orderBySaleryThenByName = orderBySaleryThenByName.thenComparing(a->a.Name);

        Stream<Author> authorStream = authors.stream().sorted(orderBySaleryThenByName);
        authors = authorStream.collect(Collectors.toList());

        StringBuffer sB = new StringBuffer();
        for(Author author: authors)
            sB.append(author.Name + " -> " + String.format("%.2f",author.getSallery())+"\n");

        return sB.toString();
    }
}
